src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureDialog.kt

```diff
@@ -0,0 +1,389 @@
+        val fragment = createTypeCodeFragment(myMethod.function, myMethod.function.retType?.typeReference)
+        val document = fragment.document!!
+        document.addDocumentListener(object : DocumentListener {
+            override fun documentChanged(event: DocumentEvent) {
+                PsiDocumentManager.getInstance(project).commitDocument(document)
+            }
+        })
+        return fragment
+    }
+
+    override fun createCallerChooser(
+        title: String?,
+        treeToReuse: Tree?,
+        callback: Consumer<MutableSet<RsFunction>>?
+    ): CallerChooserBase<RsFunction>? = null
+
+    override fun validateAndCommitData(): String? = validateAndUpdateData()
+    override fun areButtonsValid(): Boolean = isValid
+
+    override fun updateSignature() {
+        updateState()
+        super.updateSignature()
+    }
+    override fun updateSignatureAlarmFired() {
+        super.updateSignatureAlarmFired()
+        validateButtons()
+    }
+
+    override fun canRun() {
+        val error = validateAndUpdateData()
+        if (error != null) {
+            throw ConfigurationException(error)
+        }
+
+        super.canRun()
+    }
+
+    /**
+     * Updates the config from UI elements that are not updated automatically and also the validity state of the dialog.
+     */
+    private fun updateState() {
+        isValid = validateAndUpdateData() == null
+    }
+
+    private fun validateAndUpdateData(): String? {
+        val factory = RsPsiFactory(config.function.project)
+
+        if (myNameField != null) {
+            val functionName = myNameField.text
+            if (validateName(functionName)) {
+                config.name = functionName
+            } else return "Function name must be a valid Rust identifier"
+        }
+
+        if (myReturnTypeField != null) {
+            val returnTypeText = myReturnTypeField.text
+            val returnType = if (returnTypeText.isBlank()) {
+                factory.createType("()")
+            } else {
+                (myReturnTypeCodeFragment as? RsTypeReferenceCodeFragment)?.typeReference
+            }
+            if (returnType != null) {
+                config.returnTypeDisplay = returnType
+            } else {
+                return "Function return type must be a valid Rust type"
+            }
+        }
+
+        val visField = visibilityComboBox
+        if (visField != null) {
+            if (visField.hasValidVisibility) {
+                config.visibility = visField.visibility
+            } else {
+                return "Function visibility must be a valid visibility specifier"
+            }
+        }
+
+        for ((index, parameter) in config.parameters.withIndex()) {
+            if (factory.tryCreatePat(parameter.patText) == null) {
+                return "Parameter $index has invalid pattern"
+            }
+            if (parameter.typeText == null) {
+                return "Parameter $index has invalid type"
```

Comment:
Hmm, I'm not sure what to do about the second problem. The update function is only fired after you focus out of the type code fragment or after you write some character (then it's updated with the previous value).

Reply:
> The update function is only fired after you focus out of the type code fragment or after you write some character (then it's updated with the previous value)

Maybe when validating we should directly use code fragment text (instead of our variable which gets updated in callback)

Reply:
I'm using `ParameterTableModelBase.TypeColumn`, which requires a code fragment as a value. I can also try creating a new column type from scratch.

Reply:
Oh, you're right. I also had to change it for the signature, but it works. Thanks.